### ISSUE TITLE
Race condition in FiletypeImages module (inherited from FiletypeMulti)

### ISSUE THREAD

**Issue by daniel-fahl** [processwire/processwire-issues#1378]
### Short description of the issue

When AJAX uploading multiple images via AJAX upload, a race condition appears that leads to only 1-2 images being saved. It occurs in the method `\ProcessWire\FieldtypeMulti::___savePageField` where all values for a field are read, then all entries for the current page are cleared from the field's table and all values are re-INSERTed. With multiple uploads, they all read the same previous entries and then override one another.

This also leads to orphaned files in the file system, as you can see from the screenshots below.

The db transaction used is not sufficient, as it does not include reading the values.

### Expected behavior

All images should processed and saved correctly, with all images getting a corresponding entry in the field table.

### Actual behavior

When uploading multiple images at once, subsequent images override the previously uploaded ones, leading to only a few being saved despite the frontend showing successful uploads.

### Screenshots/Links that demonstrate the issue

The following screenshots illustrate my findings.

On the left I selected the images in my PW admin. On the right you see the files in the folder with the uploaded files `site/assets/files/18876` as well as the database, that I refreshed a couple times during upload.
As you see there are two entries, that override each other several times.

End result is:
The frontend 

Before:
![Screenshot1](https://user-images.githubusercontent.com/13399212/117316113-95fb6d00-ae88-11eb-9521-a838c658ad41.png)

After upload: (AJAX looks correct, DB is wrong)
![Screenshot2](https://user-images.githubusercontent.com/13399212/117316494-f25e8c80-ae88-11eb-9a71-e864cc9b2ef9.png)

After save: (UI reflects DB, several files are orphaned)
![Screenshot3](https://user-images.githubusercontent.com/13399212/117316480-ef639c00-ae88-11eb-8868-29d990cef1a5.png)

### Suggestion for a possible fix

Potentially rewrite the FiletypeMulti module so that only a delta of the field entries that have changed is saved. That way the clear of all entries for the page wouldn't affect other entries.

Another possible approach could be to limit the upload in the frontend via JS to a) avoid parallel uploads or b) calculate the sorting order for each file before the first upload, making the save easier

### Setup/Environment
<strong>Server Details</strong>

|Software|Version|
|------:|:-------|
|ProcessWire|3.0.161|
|PHP|7.2.24-0ubuntu0.18.04.7|
|Webserver|Apache|
|MySQL Server|5.7.33-0ubuntu0.18.04.1|
|MySQL Client|mysqlnd 5.0.12-dev - 20150407 - $Id: 3591daad22de08524295e1bd073aceeff11e6579 $|
<details><summary><strong>Server Settings</strong></summary>

|Parameter|Value|
|------:|:-------|
|allow_url_fopen|1|
|max_execution_time|120 (changeable)|
|max_input_nesting_level|64|
|max_input_time|60|
|max_input_vars|200000|
|memory_limit|-1|
|post_max_size|16G|
|upload_max_filesize|10G|
|xdebug||
|xdebug.max_nesting_level||
|mod_rewrite|1|
|mod_security|*confirmed off|
|EXIF Support|1|
|FreeType|1|
</details><details><summary><strong>GD Settings</strong></summary>

|Parameter|Value|
|------:|:-------|
|Version|2.2.5|
|GIF|1|
|JPG|1|
|PNG|1|
|WebP|1|
</details><details><summary><strong>iMagick Settings</strong></summary>

|Parameter|Value|
|------:|:-------|
|Version|6.9.7|
|GIF|1|
|JPG|1|
|PNG|1|
|SVG|1|
|PDF|1|
|WebP||
</details><details><summary><strong>Module Details</strong></summary>

|Module ClassName|Version|
|------:|:-------|
|AccessController|0.2.0|
|AdminHelp|1.0.9|
|AdminOnSteroids|2.0.21|
|AdminPageFieldEditLinks|3.1.2|
|AdminThemeBoss|0.6.3|
|AlgoliaSearch|0.1.0|
|BackendFormsHelper|0.0.1|
|BatchChildEditor|1.8.22|
|BreadcrumbDropdowns|0.1.22|
|CacheABTestHelper|0.0.1|
|CacheHelper|0.2.0|
|ConnectPageFields|0.3.0|
|CronjobDatabaseBackup|1.2.4|
|CustomExportCSV|1.0.0|
|CustomImportPagesCSV|2.0.0|
|CustomInputfieldDependencies|0.2.1|
|CustomRestApi|0.0.5|
|CustomWireMailMandrill|0.1.0|
|FeaturedProvider|0.0.1|
|FieldtypeAssistedURL|1.0.0|
|FieldtypeFieldsetGroup|0.0.4|
|FieldtypeFunctional|0.0.2|
|FieldtypeMultiplier|0.1.3|
|FieldtypeRepeaterMatrix|0.0.5|
|FieldtypeRuntimeMarkup|0.0.6|
|FieldtypeTable|0.1.9|
|FieldtypeTemplates|1.0.0|
|FieldtypeTextareas|0.0.7|
|FormBuilder|0.3.4|
|FrontendFormHelper|0.0.1|
|FrontendRenderingHelper|0.0.1|
|HideUneditablePages|0.0.1|
|InputfieldAssistedURL|1.0.1|
|InputfieldChosenSelect|1.2.5|
|InputfieldChosenSelectMultiple|1.2.2|
|InputfieldFormBuilderFile|0.0.2|
|InputfieldFunctional|0.0.2|
|InputfieldMultiplier|0.1.1|
|InputfieldRepeaterMatrix|0.0.5|
|InputfieldRuntimeMarkup|0.0.6|
|InputfieldTable|0.1.9|
|InputfieldTextareas|0.0.7|
|InviteListProvider|0.2.0|
|InviteProcessProvider|0.0.1|
|InviteProvider|0.2.0|
|InviteRequestProvider|0.0.1|
|LanguageHelper|0.0.2|
|LinkRedirectHelper|0.0.1|
|LoginRegister|0.0.2|
|LoginWithEmail|0.0.3|
|Migrations|0.3.1|
|MinimalFieldset|0.1.5|
|ModuleDependencies|0.2.8|
|ModuleSettingsImportExport|0.2.9|
|Oauth2LoginProvider|0.1.1|
|PageConfigProvider|0.0.1|
|PageImagesProvider|0.0.1|
|PageReviewBoss|0.2.0|
|PageTitlePreviewHelper|0.0.1|
|PageimageRemoveVariations|0.0.4|
|ProDrafts|0.0.8|
|ProcessAdminHelp|1.0.8|
|ProcessChildrenCsvExport|1.8.21|
|ProcessForgotPassword|1.0.2|
|ProcessFormBuilder|0.3.4|
|ProcessMigrations|0.1.3|
|ProcessPageListerPro|1.1.2|
|ProcessProDrafts|0.0.8|
|ProcessSetupPageName|2.1.4|
|ProcessWireUpgrade|0.0.7|
|ProcessWireUpgradeCheck|0.0.7|
|RelatedProvider|0.1.0|
|SearchProvider|0.1.2|
|TemplateCategoryEntry|0.0.1|
|TemplateChurchEntry|0.0.1|
|TemplateCollectionEntry|0.0.1|
|TemplateDenominationEntry|0.0.1|
|TemplateFieldWidths|0.1.10|
|TemplateInviteEntry|0.2.0|
|TemplateMailHelper|0.0.1|
|TemplatePage|0.0.1|
|TemplateResourceContentEntry|0.0.1|
|TemplateResourceContentEntryHelper|0.0.1|
|TemplateResourceEntry|0.0.1|
|TemplateResourceEntryHelper|0.0.1|
|TemplateUser|0.0.2|
|TextformatterMakeLinks|1.0.1|
|TextformatterOEmbed|2.0.2|
|TextformatterTagParser|2.2.0|
|TextformatterVideoEmbedOptions|0.3.6|
|TimedPublishHelper|0.0.1|
|TracyDebugger|4.21.19|
|UserAdminProvider|0.0.1|
|UserMailNotifications|0.0.1|
</details>

**Comment by adrianbj** [comment]
@daniel-fahl - any chance that this commit (https://github.com/processwire/processwire/commit/644f0fb3fd8c6f97b531ae3c1a8f4d883edb6b3a) from yesterday fixes this?

**Comment by adrianbj** [comment]
@daniel-fahl - before you try that commit, please note this issue I just discovered (which may or may not affect your use)
https://github.com/processwire/processwire/commit/644f0fb3fd8c6f97b531ae3c1a8f4d883edb6b3a#commitcomment-50475685

**Comment by ryancramerdesign** [comment]
@daniel-fahl I can't duplicate this one. I've got a client currently uploading hundreds of images a day in large batches, and there haven't been any reports similar to this. I'm wondering if there might be another factor at play. Your screenshots show an admin environment that looks like it may have a lot of modifications to it. Can you try with those disabled and using the standard admin theme (AdminThemeUikit) just in case? Also I'd recommend upgrading your PW version to the current dev or at least the current master since you are running an old version... though I'm not aware of this particular issue in a prior version, but good to upgrade just in case.

@adrianbj Those code updates won't affect this particular action as I don't think FieldtypeFile/FieldtypeImage uses the savePageFieldRows() method.

**Comment by daniel-fahl** [comment]
Thanks @ryancramerdesign and @adrianbj 

I just disabled any admin related modules I could find, and reproduced the exact same behaviour. Attached another screenshot that again shows that entries in the database are overridden by concurrent uploads.

![image](https://user-images.githubusercontent.com/13399212/117791579-c5282a80-b24a-11eb-8e61-fc797176ae62.png)

**Comment by teppokoivula** [comment]
This is likely unrelated, but mentioning just in case: I've seen similar issues happen in an environment behind load balancer. Can't remember the specifics, but I believe it was somehow related to subsequent requests going to separate servers and sessions. Not sure if this helps much, but might be good to rule out any server setup related issues anyway.

**Comment by daniel-fahl** [comment]
Thanks @teppokoivula - I'll take a closer look into that.
It's most likely not the main issue though, since I can reproduce it both on our production environment as well as my local machine (Apache 2.4.46, MySQL 5.7.32 and PHP 7.2.34 NTS all running natively on MacOS 11.2.3), without any load balancing or http caching afaik

**Comment by daniel-fahl** [comment]
@ryancramerdesign I just set up a clean installation of PW on my machine, and observed the same issue as described above.
After installation I only created a subpage to the "about" page and uploaded images into the already present `images` field.

You can see the results in my screenshot.


![Screenshot01](https://user-images.githubusercontent.com/13399212/117825495-a9834b00-b26f-11eb-9685-d6df51dc070d.png)
![Screenshot02](https://user-images.githubusercontent.com/13399212/117825477-a720f100-b26f-11eb-9a18-347ff6a07492.png)

**Comment by daniel-fahl** [comment]
@ryancramerdesign do you have an ETA for looking into this?
Feel free to contact me if you require any additional information!

**Comment by ryancramerdesign** [comment]
@daniel-fahl Thus far I can't duplicate it here, so I think we need someone else that can duplicate it before it's clear how to proceed. Multiple image uploads of this sort are so common that I think if the issue were easy to reproduce, we'd be hearing about it every day. So I think there must be some other factor present here, but we just need more people experiencing the issue before we can identify what they have in common. 

@teppokoivula On a load balancer, any requests that upload files have to upload them to the node that others mirror from (the primary node). Or just target all admin requests to the primary node. That's what we're doing here at least.

**Comment by teppokoivula** [comment]
> @teppokoivula On a load balancer, any requests that upload files have to upload them to the node that others mirror from (the primary node). Or just target all admin requests to the primary node. That's what we're doing here at least.

Just to be clear: this was not an issue with files themselves, but rather sessions, and more specifically the way core keeps track of temp files (in Pagefiles). One server instance was overwriting the session variable set by another, which meant that the files were considered "temporary" and removed.

Anyway, while the result was similar to that described here, at this point I'm pretty sure this has nothing to do with the issue covered by this report :)

**Comment by BitPoet** [comment]
A bit late to the game, but since I just stumbled about this while looking for something else: could this issue have the same underlying reason as #798?

**Comment by daniel-fahl** [comment]
Thanks @BitPoet - it indeed looks like it might be connected. We previously switched to storing native php sessions in Redis, so this might be what triggers this issue. (It also feels like a race condition from what is discussed over there...)

**Comment by ryancramerdesign** [comment]
@daniel-fahl @BitPoet Did you try the solution mentioned on that #798 ? having session locking enabled would be necessary, otherwise file uploads would crash into each other and some files would be missing. So if session locking is somehow disabled then that would definitely result in the issue discussed here, not to mention others, as sessions could corrupt each other.

**Comment by matjazpotocnik** [comment]
@daniel-fahl Any news? Can this issue be closed?

**Comment by daniel-fahl** [comment]
Sorry for not responding earlier.
I tried implementing the solution from #798 but the redis config keys seem to have changed. Due to several shifts in priority this project hasn't received the attention it should lately.
I'll try to find out the correct redis config for our setup and give an update here asap

**Comment by daniel-fahl** [comment]
Okay, wow.
I am very sorry for those delays, but I finally managed to find the source of my problem, and I even fixed it!

When trying the configuration mentioned in #798 initially, it didn't change anything. I got suspicious when I noticed the variables `redis.session.*` were neither shown in `phpinfo()` nor via `ini_get_all()`.

After too much research and a lot of learning I finally found these variables in the phpredis repository, and noticed they appeared in the README.md and source code starting from phpredis version 5.0.0. After learning how to find out which version my installed php extension is (`php --ri redis` in case anyone finding this is also wondering) I noticed the default Ubuntu 18.4 apt repository shipped me with version `3.1.6-1build1` - so the variables weren't even considered. It seems session locking only found its way into the extension later on.

Long story short, after uninstalling the apt version and installing the (5.3.6) PECL version (we're still on 7.2 unfortunately) Redis now correctly uses session locking with the following settings: (taken with slight adjustment from the phpredis docs)

```ini
; Should the locking be enabled? Defaults to: 0.
redis.session.locking_enabled = 1
; How long should the lock live (in seconds)? Defaults to: value of max_execution_time.
redis.session.lock_expire = 60
; How long to wait between attempts to acquire lock, in microseconds (µs)?. Defaults to: 2000
redis.session.lock_wait_time = 50000
; Maximum number of times to retry (-1 means infinite). Defaults to: 10
redis.session.lock_retries = -1
```

I really hope someone stumbles across this when facing a similar issue.


@ryancramerdesign in fact, until recently phpredis had rather poor defaults for the `redis.session.*` ini settings with session locking being **disabled by default**! As it appears from [this merge request](https://github.com/phpredis/phpredis/pull/1908) those should be a lot more sane starting with the latest or next version of the extension - which is little comfort for people like me (until now) depending on old os repository binaries.
